File: leetcode/python/0415_add_strings.py
class Solution:
    def addStrings(self, num1: str, num2: str) -> str:
        # loop through num1 and num2 using i and j respectively:
        # if i < 0: n1 = 0
        # else: n1 = int(num1[i])
        # if j < 0: n1 = 0
        # else: n2 = int(num2[i])
        # if carry == 0:
        #   if n1 + n2 < 10: res += str((n1 + n2) % 10)
        #   elif  n1 + n2 >= 10: res += str((n1 + n2) % 10), carry = 1
        # if carry == 1:
        #   if carry + n1 + n2 < 10: res += str((carry + n1 + n2) % 10), carry = 0
        #   elif carry + n1 + n2 >= 10: res += str((carry + n1 + n2) % 10)
        # i -= 1, j -= 1
        i, j, carry, res = len(num1) - 1, len(num2) - 1, 0, ""

        while i >= 0 or j >= 0:
            n1, n2 = 0, 0
            if i >= 0: n1 = int(num1[i])
            if j >= 0: n2 = int(num2[j])

            if carry == 0:
                if n1 + n2 < 10: res += str((n1 + n2) % 10)
                elif n1 + n2 >= 10: 
                    res += str((n1 + n2) % 10)
                    carry = 1
            elif carry == 1:
                if carry + n1 + n2 < 10:
                    res += str((carry + n1 + n2) % 10)
                    carry = 0
                elif carry + n1 + n2 >= 10:
                    res += str((carry + n1 + n2) % 10)
            
            i -= 1
            j -= 1

        if carry == 1: res += "1"

        return res[::-1]

File: leetcode/java/0415_add_strings.java
class Solution {
    public String addStrings(String num1, String num2) {
        StringBuilder res = new StringBuilder();
        int len1 = num1.length(), len2 = num2.length();
        int val1 = 0, val2 = 0, carry = 0;

        for (int i = len1 - 1, j = len2 - 1; i >= 0 || j >= 0; i--, j--) {
            if (i < 0) val1 = 0;
            else if (i >= 0) val1 = Integer.valueOf(String.valueOf(num1.charAt(i)));

            if (j < 0) val2 = 0;
            else if (j >= 0) val2 = Integer.valueOf(String.valueOf(num2.charAt(j)));

            int val = (val1 + val2 + carry) % 10;

            if (carry == 0) {
                if (val1 + val2 > 9) carry = 1;
                res.append(val);
            } else if (carry == 1) {
                if (val1 + val2 + carry < 10) carry = 0;
                res.append(val);
            }
        }

        if (carry == 1) res.append("1");
        return res.reverse().toString();
    }
}